src/com/ms/silverking/cloud/topology/TopologyParser.java

```diff
@@ -11,157 +11,158 @@
+                  break;
+                default:
+                  throw new RuntimeException("panic");
+                }
+              } else {
+                switch (token.getType()) {
+                case OPEN_BLOCK:
+                  if (prevToken != null) {
+                    switch (prevToken.getType()) {
+                    case OPEN_BLOCK:
+                    case CLOSE_BLOCK:
+                      throw new TopologyParseException("Invalid block defiition");
+                    case ENTRY:
+                      TopologyEntry parent;
+                      TopologyEntry lastAdded;
+
+                      parent = parentEntries.get(parentEntries.size() - 1);
+                      lastAdded = parent.getLastChild();
+                      parentEntries.add(lastAdded);
+                      break;
+                    default:
+                      throw new RuntimeException("panic");
                     }
+                  }
+                  break;
+                case CLOSE_BLOCK:
+                  switch (prevToken.getType()) {
+                  case CLOSE_BLOCK:
+                  case OPEN_BLOCK:
+                  case ENTRY:
+                    parentEntries.remove(parentEntries.size() - 1);
+                    break;
+                  default:
+                    throw new RuntimeException("panic");
+                  }
+                  break;
+                case ENTRY:
+                  parentEntries.get(parentEntries.size() - 1).addChild(new TopologyEntry(token.getEntry()));
+                  break;
+                default:
+                  throw new RuntimeException("panic");
                 }
-            } while (line != null);
-            if (debug) {
-                System.out.println(root);
-            }
-            rootNode = TopologyEntry.entriesToNodes(root);
-            if (debug) {
-                System.out.println(rootNode.toStructuredString());
-            }
-            if (version == noVersion) {
-                return Topology.fromRoot(null, rootNode);
-            } else {
-                return Topology.fromRoot(null, rootNode, version);
+                prevToken = token;
+              }
             }
-        } catch (TopologyParseException tpe) {
-            throw new TopologyParseException(tpe.getMessage() +" lineNumber: "+ lineNumber);
-        } catch (IOException ioe) {
-            throw new IOException(ioe.getMessage() +" lineNumber: "+ lineNumber);
+          }
         }
+      } while (line != null);
+      if (debug) {
+        System.out.println(root);
+      }
+      rootNode = TopologyEntry.entriesToNodes(root);
+      if (debug) {
+        System.out.println(rootNode.toStructuredString());
+      }
+      if (version == noVersion) {
+        return Topology.fromRoot(null, rootNode);
+      } else {
+        return Topology.fromRoot(null, rootNode, version);
+      }
+    } catch (TopologyParseException tpe) {
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

**Problem**: While wrapping the caught exception into a custom one, information about the caught exception is being lost, including information about the stack trace of the exception.

**Fix**: If the caught exception object does not contain sensitive information, consider passing it as the &#34;rootCause&#34; or inner exception parameter to the constructor of the new exception before throwing the new exception. (Note that not all exception constructors support inner exceptions. Use a wrapper exception that supports inner exceptions.) 
[Learn more](https://www.ibm.com/support/pages/best-practice-catching-and-re-throwing-java-exceptions)
Similar issue at line numbers 124.